File: random-forest/r/random_forest.R
# Use SigOpt to tune a Random Forest Classifier in R
# Learn more about SigOpt's R Client:
# https://sigopt.com/docs/overview/r

# Install packages
install.packages("devtools", repos = "http://cran.us.r-project.org")
library(devtools)
install_github("sigopt/SigOptR")
library(SigOptR)
install.packages('randomForest', repos = "http://cran.us.r-project.org")
library(randomForest)


# Learn more about authenticating the SigOpt API:
# https://sigopt.com/docs/overview/authentication
Sys.setenv(SIGOPT_API_TOKEN="INSERT_YOUR_TOKEN_HERE")


# Load dataset
# We are using the iris dataset as an example
library(datasets)
X <- subset(iris, select = -c(Species))
y <- iris$Species

# Create a SigOpt experiment for the Random Forest parameters
experiment <- create_experiment(list(
  name="Random Forest (R)",
  parameters=list(
    list(name="mtry", type="int", bounds=list(min=1, max=ncol(iris)-1)),
    list(name="ntree", type="int", bounds=list(min=1, max=100)),
    list(name="sampsize", type="double", bounds=list(min=0.25, max=1.0)),
    list(name="nodesize", type="int", bounds=list(min=1, max=10))
  )
))
print(paste("Created experiment: https://sigopt.com/experiment", experiment$id, sep="/"))

# Our object metric is the mean of cross validated accuracies
# We use cross validation to prevent overfitting
evaluate_model <- function(assignments, X, y) {
  # evaluate cross folds for accuracy
  num_folds = 5
  cv_accuracies = c()
  for (k in 1:num_folds){
    cv_split <- sample(2, nrow(iris), replace=TRUE, prob=c(1-1.0/num_folds, 1.0/num_folds))
    X_train <- X[cv_split==1, ]
    y_train <- y[cv_split==1]
    X_valid <- X[cv_split==2, ]
    y_valid <- y[cv_split==2]

    rf_fit <- randomForest(y=y_train,
                           x=X_train,
                           ntree=assignments$ntree,
                           mtry=assignments$mtry,
                           sampsize=assignments$sampsize*nrow(X_train),
                           nodesize=assignments$nodesize,
                           proximity=TRUE)
    prediction <- predict(rf_fit, X_valid)
    correct_predictions <- prediction == y_valid
    accuracy <- sum(correct_predictions)/nrow(X_valid)
    cv_accuracies <- c(accuracy, cv_accuracies)
  }
  return(c(mean(cv_accuracies), sd(cv_accuracies)))
}

# Run the Optimization Loop between 10x - 20x the number of parameters
for (i in 1:80) {
  # Receive a Suggestion from SigOpt
  suggestion <- create_suggestion(experiment$id)

  # Evaluate the model locally
  res <- evaluate_model(suggestion$assignments, X, y)

  # Report an Observation (with standard deviation) back to SigOpt
  create_observation(experiment$id, list(suggestion=suggestion$id,
                                         value=res[1],
                                         value_stddev=res[2]))
}

# Re-fetch the experiment to get the best observed value and assignments
experiment <- fetch_experiment(experiment$id)
best_assignments <- experiment$progress$best_observation$assignments

# To wrap up the Experiment, fit the RandomForest on the best assigments
# and train on all available data
rf <- randomForest(x=X,
                   y=y,
                   ntree=best_assignments$ntree,
                   mtry=best_assignments$mtry,
                   sampsize=best_assignments$sampsize*nrow(X),
                   nodesize=best_assignments$nodesize,
                   proximity=TRUE)

File: random-forest/python/random_forest.py
# Use SigOpt to tune a Random Forest Classifier in Python
# Learn more about SigOpt's Python Client:
# https://sigopt.com/docs/overview/python

# Run `pip install sigopt` to download the python API client
# Run `pip install sklearn` to install scikit-learn, a machine learning
# library in Python (http://scikit-learn.org)
from sigopt import Connection
from sklearn import datasets
from sklearn.model_selection import cross_val_score, ShuffleSplit
from sklearn.ensemble import RandomForestClassifier
import numpy

# Learn more about authenticating the SigOpt API:
# https://sigopt.com/docs/overview/authentication
conn = Connection(client_token=SIGOPT_API_TOKEN)

# Load dataset
# We are using the iris dataset as an example
iris = datasets.load_iris()
X = iris.data
y = iris.target

# Create a SigOpt experiment for the Random Forest parameters
experiment = conn.experiments().create(
  name="Random Forest (Python)",
  project="sigopt-examples",
  metrics=[dict(name='accuracy', objective='maximize')],
  parameters=[
    dict(name="max_features", type="int", bounds=dict(min=1, max=len(iris)-2)),
    dict(name="n_estimators", type="int", bounds=dict(min=1, max=100)),
    dict(name="min_samples_leaf", type="int", bounds=dict(min=1, max=10))
  ]
)
print("Created experiment: https://sigopt.com/experiment/" + experiment.id)

# Our object metric is the mean of cross validated accuracies
# We use cross validation to prevent overfitting
def evaluate_model(assignments, X, y):
  # evaluate cross folds for accuracy
  cv = ShuffleSplit(
    n_splits=5,
    test_size=0.3,
  )
  classifier = RandomForestClassifier(
    n_estimators=assignments['n_estimators'],
    max_features=assignments['max_features'],
    min_samples_leaf=assignments['min_samples_leaf']
  )
  cv_accuracies = cross_val_score(classifier, X, y, cv=cv)
  return (numpy.mean(cv_accuracies), numpy.std(cv_accuracies))

# Run the Optimization Loop between 10x - 20x the number of parameters
for _ in range(60):
    # Receive a Suggestion from SigOpt
    suggestion = conn.experiments(experiment.id).suggestions().create()

    # Evaluate the model locally
    (value, std) = evaluate_model(suggestion.assignments, X, y)

    # Report an Observation (with standard deviation) back to SigOpt
    conn.experiments(experiment.id).observations().create(
        suggestion=suggestion.id,
        value=value,
        value_stddev=std,
    )

# Re-fetch the best observed value and assignments
best_assignments = conn.experiments(experiment.id).best_assignments().fetch().data[0].assignments

# To wrap up the Experiment, fit the RandomForest on the best assignments
# and train on all available data
rf = RandomForestClassifier(
    n_estimators=best_assignments['n_estimators'],
    max_features=best_assignments['max_features'],
    min_samples_leaf=best_assignments['min_samples_leaf']
)
rf.fit(X, y)
